doc/decisions/DR016-Related-entities-result-paging.md

```diff
@@ -0,0 +1,352 @@
+similar pros and cons, but that option was discarded (see Background
+section).
+
+```python
+class AbstractPageToken:
+    ...
+
+# Callback taking index of batch element and list of related entities.
+SuccessCallback = Callable[[int, List[EntityReferences]], None]
+
+def relatedEntities(
+        batch: List[EntityReference],
+        relationship,
+        successCb: SuccessCallback,
+        errorCb,
+        limit: int = -1,
+        pageToken: AbstractPageToken = None) -> AbstractPageToken: ...
+```
+
+#### Pros
+
+- Conceptually simple.
+- Whole batch is paged at once, retaining batch-awareness in the
+  manager for subsequent pages.
+- Per-element callback means it's trivial to skip processing elements of
+  the input batch whose pages are exhausted.
+- Matches the most common pattern of pageable HTTP APIs allowing e.g.
+  the page token to be a thin wrapper around a "real" page token(s).
+- Retains existing design in line with OpenAssetIO design principles,
+  i.e. non-paged signature is simply the same function but ignoring
+  optional arguments.
+- Manager plugin can remain conceptually stateless, since required state
+  is bundled up in the page token.
+- Page token is (or can be) immutable, aiding thread-safety.
+
+#### Cons
+
+- Adding more functionality (e.g. previous page) requires additional
+  sibling functions or arguments.
+- Adding metadata (e.g. total results) requires additional sibling
+  functions or arguments.
+- Likely that the manager plugin must combine multiple back-end page
+  tokens into one, adding complexity to ensure they are maintained in
+  sync.
+- Batch elements cannot enter an error state in isolation, since paging
+  is whole-batch.
+
+
+### Option 2 - Cursor objects / iterators
+
+Firstly let's consider the general pros and cons of object-oriented
+cursor objects.
+
+```python
+class AbstractPage:
+
+    def get(self) -> List[EntityReferences]:
+        # Current page of refs.
+        ...
+
+    def next(self) -> AbstractPage:
+        # Get next page.
+        ...
+
+    def remaining(self) -> Union[int, None]:
+        # None means "unsupported".
+        ...
+```
+
+#### Pros
+
+- Retrieving the data can be decoupled from traversal of pages, allowing
+  for lazy network retrieval of desired pages.
+- Additional metadata can be included in the cursor (e.g. remaining
+  pages) in an extensible way.
+- A mutable, stateful cursor object maximizes opportunities for the
+  manager plugin to optimize query caches.
+
+#### Cons
+
+- Manager plugin must implement yet another interface.
+- A mutable, stateful cursor violates the stateless design principle of
+  interactions with manager plugins. This causes complications, e.g.
+  thread-safety.
+- An immutable cursor object (such that `.next()` returns a new cursor
+  object), could increase optimisation complexity for the manager
+  implementation, e.g. managing query cache sharing, lifetime and
+  updates across multiple "live" cursors.
+
```

Comment:
[fix] - Cursor objects increase complexity when binding to multiple target languages.

Reply:
Added con in 1003d8be